moonraker/components/filament_manager.py

```diff
@@ -0,0 +1,296 @@
+        return self.db.get(ACTIVE_SPOOL_KEY, None)
+
+    def add_spool(self, data: {}) -> str:
+        spool = Spool(data)
+        missing_attrs = spool.validate()
+        if missing_attrs:
+            raise self.server.error(
+                f"Missing spool attributes: {missing_attrs}", 404)
+
+        next_spool_id = 0
+        spools = self.db.get(SPOOLS_SUB_NS, [])
+        if spools:
+            next_spool_id = int(list(spools)[-1], 16) + 1
+        spool_id = f"{next_spool_id:06X}"
+
+        self.db.insert(self._spool_id(spool_id), spool.serialize())
+        logging.info(f'New spool added, id: {spool_id}')
+
+        return spool_id
+
+    def update_spool(self, spool_id: str, data: {}) -> None:
+        spool_from_db = self.find_spool(spool_id)
+
+        spool = Spool(spool_from_db)
+        spool.update(data)
+        missing_attrs = spool.validate()
+        if missing_attrs:
+            raise self.server.error(
+                f"Missing spool attributes: {missing_attrs}", 404)
+
+        self.db.update_child(self._spool_id(spool_id), spool.serialize())
+        logging.info(f'Spool id: {spool_id} updated.')
+
+        return
+
+    def delete_spool(self, spool_id: str) -> None:
+        self.db.delete(self._spool_id(spool_id))
+        logging.info(f'Spool id: {spool_id} deleted.')
+        return
+
+    def find_all_spools(self) -> List[dict]:
+        return self.db.get(SPOOLS_SUB_NS, [])
+
+    def track_filament_usage(self, job_id: str, used_length: float):
+        spool_id = self.get_active_spool_id()
+        spool = self.find_spool(spool_id)
+
+        if spool:
+            old_used_length = spool['used_length']
+            new_used_length = old_used_length + used_length
+            spool['used_length'] = new_used_length
+
+            diameter = spool['diameter']
+            density = spool['density']
+            old_used_weight = spool['used_weight']
+            cost = spool['cost']
+            total_weight = spool['total_weight']
+            used_weight = 0
+            new_used_weight = 0
+            used_cost = 0
+
+            if diameter and density:
+                r = diameter/2
+                density_mm = density/1000
+                used_weight = math.pi * r * r * used_length * density_mm
+                new_used_weight = old_used_weight + used_weight
+                spool['used_weight'] = new_used_weight
+
+            if cost and used_weight and total_weight:
+                used_cost = used_weight/total_weight*cost
+
+            if not spool['first_used']:
+                spool['first_used'] = time.time()
+
+            spool['last_used'] = time.time()
+
+            self.update_spool(spool_id, spool)
+
+            history = self.server.lookup_component('history', None)
+            metadata = {'spool_id': spool_id,
+                        'used_weight': used_weight,
+                        'cost': used_cost}
+            history.add_job_metadata(job_id, {'filament': metadata})
```

Comment:
Maybe spool_manager can publish event which can be consumed by history component?

Reply:
@Arksine is there any reason why the lookup_component method doesn't return a None value if the component isn't valid so modules can gracefully check (and reduce functionality) or fail (if it's required) if the component isn't found?

A spool change notification would be good either way as it would notify front ends that they should update the active spool info (which I'm assuming will be shown).

I believe we should also allow for the case of multiple spools used during one print as changing colors/filaments mid-print seems to be a common thing (especially with the pause at layer feature in slicers). Which means you'd need to make a spools list in metadata rather than a spool dict. Spools would also have to be attached to jobs mid-print.

Reply:
I think in an ideal scenario the `history` module will look up `spool_manager` and add a reference to the spool if it has been configured.  The `spool_manager` can emit a "spool_changed" event that that `history` can register.  When a spool change is noted then `history` can add the current spool's reference to the ongoing job.